core/src/multipart_form_data.cpp

```diff
@@ -0,0 +1,179 @@
+#include "irods/private/http_api/multipart_form_data.hpp"
+
+#include "irods/private/http_api/log.hpp"
+
+#include <boost/algorithm/string.hpp>
+#include <boost/beast/http/rfc7230.hpp>
+
+#include <fmt/format.h>
+
+namespace irods::http
+{
+	auto get_multipart_form_data_boundary(const std::string_view _data) -> std::optional<std::string_view>
+	{
+		// The ext_list type does not treat "multipart/form-data" as a single string.
+		// I'm not sure if this is by design, but we have to skip the multipart/ prefix so that
+		// ext_list can extract the boundary argument.
+		//
+		// TODO Consider opening an issue for Boost.Beast in Github.
+
+		const auto pos = _data.find("form-data");
+
+		if (pos == std::string_view::npos) {
+			return std::nullopt;
+		}
+
+		boost::beast::http::ext_list list{_data.substr(pos)};
+		const auto iter = list.find("form-data");
+
+		if (iter == std::end(list)) {
+			log::error("{}: Missing [boundary] parameter in [multipart/form-data] value.", __func__);
+			return std::nullopt;
+		}
+
+		for (auto&& param : iter->second) {
+			if (param.first == "boundary") {
+				log::debug("PARAM = (name={}, value={})", param.first, param.second);
+				return param.second;
+			}
+		}
+
+		return std::nullopt;
+	} // get_multipart_form_data_boundary
+
+	auto parse_multipart_form_data(const std::string_view _boundary, const std::string_view _data)
+		-> query_arguments_type
+	{
+		if (_data.empty()) {
+			return {};
+		}
+
+		const auto boundary_start = fmt::format("--{}", _boundary);
+		const auto boundary_end = fmt::format("{}--", boundary_start);
+
+		enum class parser_state
+		{
+			find_boundary_start,
+			read_boundary_start,
+			read_header,
+			read_body
+		};
+
+		query_arguments_type args;
+		auto pstate = parser_state::find_boundary_start;
+		std::string_view::size_type pos = 0;
+		std::string param_name;
+
+		while (true) {
+			switch (pstate) {
+				case parser_state::find_boundary_start: {
+					auto tpos = _data.find(boundary_start, pos);
+					if (std::string_view::npos == tpos) {
+						log::error("expected boundary start [{}]. malfored message structure.", boundary_start);
+						return args;
+					}
+					pos = tpos;
+
+					// Did we actually find a boundary end marker?
+					if (boundary_end == _data.substr(pos, boundary_end.size())) {
+						log::trace("found boundary end [{}]. we're done.", boundary_end);
+						return args;
+					}
+
+					// We found a boundary start marker.
+					pstate = parser_state::read_boundary_start;
+					break;
+				}
+
+				case parser_state::read_boundary_start: {
+					// Check that the boundary start ends with a CRLF sequence.
+					const auto crlf_pos = pos + boundary_start.size();
+					if ("\r\n" != _data.substr(crlf_pos, 2)) {
+						log::error("expected CRLF [\\r\\n] after boundary start. malfored message structure.");
```

Comment:
malfored -> malformed

Saw this in many of the other messages as well, so might want do a find/replace.

Reply:
Will do.

Reply:
Done